core/src/main/java/org/elasticsearch/transport/TcpTransport.java

```diff
@@ -1053,6 +1053,23 @@ private BytesReference buildMessage(long requestId, byte status, Version nodeVer
         return new CompositeBytesReference(header, messageBody, zeroCopyBuffer);
     }
 
+    protected abstract int length(Buffer buffer);
```

Comment:
I don't think we should do this, Instead we should implement a simple wrapper around Buffer extending BytesReference, that's why we cleaned this up all over the place to be simple and easy to implement. Just throw UOE where applicable.